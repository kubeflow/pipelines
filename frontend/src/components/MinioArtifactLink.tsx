import * as React from 'react';
import { StoragePath, StorageService } from '../lib/WorkflowParser';
import { S3Artifact } from '../../third_party/argo-ui/argo_template';
import { buildQuery } from 'src/lib/Utils';

const artifactApiUri = ({ source, bucket, key }: StoragePath, namespace?: string) =>
  `namespaces/${namespace}/artifacts/get${buildQuery({ source, bucket, key })}`;

/**
 * A component that renders an artifact link.
 */
const MinioArtifactLink: React.FC<{
  artifact: Partial<S3Artifact> | null | undefined;
  namespace?: string;
}> = ({ artifact, namespace }) => {
  if (!artifact || !artifact.key || !artifact.bucket) {
    return null;
  }

  const { key, bucket, endpoint } = artifact;
  const source = endpoint === 's3.amazonaws.com' ? StorageService.S3 : StorageService.MINIO;
  const linkText = `${source.toString()}://${bucket}/${key}`;
  // Opens in new window safely
  return (
    <a
      href={artifactApiUri({ key, bucket, source }, namespace)}
      target={'_blank'}
      rel='noreferrer noopener'
      title={linkText}
    >
      {linkText}
    </a>
  );
};

export default MinioArtifactLink;
