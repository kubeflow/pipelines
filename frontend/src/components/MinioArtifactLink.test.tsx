/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import MinioArtifactLink from './MinioArtifactLink';

describe('MinioArtifactLink', () => {

  it('handles null artifact', () => {
        expect(MinioArtifactLink(null as any)).toMatchSnapshot();
  });

  it('handles empty artifact', () => {
    expect(MinioArtifactLink({} as any)).toMatchSnapshot();
  });

  it('handles invalid artifact: no bucket', () => {
    const s3artifact = {
      accessKeySecret: {key: 'accesskey', optional: false, name: 'minio'},
      bucket: '',
      endpoint: 'minio.kubeflow',
      key: 'bar',
      secretKeySecret: {key: 'secretkey', optional: false, name: 'minio'},
    };
    expect(MinioArtifactLink(s3artifact)).toMatchSnapshot();
  });

  it('handles invalid artifact: no key', () => {
    const s3artifact = {
      accessKeySecret: {key: 'accesskey', optional: false, name: 'minio'},
      bucket: 'foo',
      endpoint: 'minio.kubeflow',
      key: '',
      secretKeySecret: {key: 'secretkey', optional: false, name: 'minio'},
    };
    expect(MinioArtifactLink(s3artifact)).toMatchSnapshot();
  });

  it('handles s3 artifact', () => {
    const s3artifact = {
      accessKeySecret: {key: 'accesskey', optional: false, name: 'minio'},
      bucket: 'foo',
      endpoint: 's3.amazonaws.com',
      key: 'bar',
      secretKeySecret: {key: 'secretkey', optional: false, name: 'minio'},
    };
    expect(MinioArtifactLink(s3artifact)).toMatchSnapshot();
  });

  it('handles minio artifact', () => {
    const minioartifact = {
      accessKeySecret: {key: 'accesskey', optional: false, name: 'minio'},
      bucket: 'foo',
      endpoint: 'minio.kubeflow',
      key: 'bar',
      secretKeySecret: {key: 'secretkey', optional: false, name: 'minio'},
    };
    expect(MinioArtifactLink(minioartifact)).toMatchSnapshot();
  });

});