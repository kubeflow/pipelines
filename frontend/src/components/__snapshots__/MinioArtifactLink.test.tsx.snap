// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MinioArtifactLink handles empty artifact 1`] = `null`;

exports[`MinioArtifactLink handles invalid artifact: no bucket 1`] = `null`;

exports[`MinioArtifactLink handles invalid artifact: no key 1`] = `null`;

exports[`MinioArtifactLink handles minio artifact 1`] = `
<a
  href="artifacts/get?source=minio&bucket=foo&key=bar"
  rel="noreferrer noopener"
  target="_blank"
  title="minio://foo/bar"
>
  minio://foo/bar
</a>
`;

exports[`MinioArtifactLink handles null artifact 1`] = `null`;

exports[`MinioArtifactLink handles s3 artifact 1`] = `
<a
  href="artifacts/get?source=s3&bucket=foo&key=bar"
  rel="noreferrer noopener"
  target="_blank"
  title="s3://foo/bar"
>
  s3://foo/bar
</a>
`;
